capsules/src/net/udp/driver.rs

```diff
@@ -261,32 +190,29 @@ impl<'a> UDPDriver<'a> {
 
             // Send UDP payload. Copy payload into packet buffer held by this driver, then queue
             // it on the udp_mux.
-            let result = app
-                .app_write
-                .as_ref()
-                .map_or(ReturnCode::ENOMEM, |payload| {
-                    self.kernel_buffer
-                        .take()
-                        .map_or(ReturnCode::ENOMEM, |mut kernel_buffer| {
-                            kernel_buffer[0..payload.len()].copy_from_slice(payload.as_ref());
-                            kernel_buffer.slice(0..payload.len());
-                            match self.sender.driver_send_to(
-                                dst_addr,
-                                dst_port,
-                                src_port,
-                                kernel_buffer,
-                                self.driver_send_cap,
-                                self.net_cap,
-                            ) {
-                                Ok(_) => ReturnCode::SUCCESS,
-                                Err(mut buf) => {
-                                    buf.reset();
-                                    self.kernel_buffer.replace(buf);
-                                    ReturnCode::FAIL
-                                }
+            let result = app.app_write.map_or(ReturnCode::ENOMEM, |payload| {
+                self.kernel_buffer
+                    .take()
+                    .map_or(ReturnCode::ENOMEM, |mut kernel_buffer| {
+                        kernel_buffer[0..payload.len()].copy_from_slice(payload.as_ref());
```

Comment:
What will happen here if the app allows a slice longer than the `kernel_buffer`? (It's limited by `max_tx_pyld_len`, is that checked to be in bounds of the `kernel_buffer` length?)

Reply:
This is a good point. I should update the porting guide on how to deal with slices/buffers of different sizes, especially when they might disappear from under you.

Reply:
This will panic if the app allows a slice longer than the `kernel_buffer`. The UDP component checks that the `kernel_buf` is longer than `max_tx_pyld_len`, so as long as the UDP capsule is constructed via the component, it should be impossible for an app to do so given the checks in `allow()`.

Actually, after checking, the UDP component mistakenly set the `kernel_buf` length to 8 bytes smaller `max_tx_pyld_len`, so I have fixed that as well.

I am confused by "they might disappear from under you" comment -- isn't this covered by the `map_or()`?

Reply:
Sorry, I was writing imprecisely.

I'm generally worried about capsule code assuming things about the length of application slices: they can change in length between invocations. In this particular case, what happens if the application slice is longer than the kernel buffer? An application could provide an arbitrary length AppSlice. I see this is checked in allow(), but this checks against `max_tx_payld_len`, not the actual length of `kernel_buf`. So someone accidentally passes the wrong value when creating the driver, then in some weird cases suddenly the kernel panics when an application sends a long buffer.

I think this code would be more robust/safter if it explicitly looked at the length of `kernel_buffer`.

Reply:
I added an extra check.